sycl/include/sycl/handler.hpp

```diff
@@ -318,45 +318,105 @@ checkValueRange(const T &V) {
 #endif
 }
 
-template <typename TransformedArgType, int Dims, typename KernelType>
-class RoundedRangeKernel {
+template <int Dims> class RoundedRangeIDGenerator {
+  id<Dims> Id;
+  range<Dims> OldRange;
+  range<Dims> NewRange;
+  bool Done = false;
+
 public:
-  RoundedRangeKernel(range<Dims> NumWorkItems, KernelType KernelFunc)
-      : NumWorkItems(NumWorkItems), KernelFunc(KernelFunc) {}
+  RoundedRangeIDGenerator(const id<Dims> &Id, const range<Dims> &OldRange,
+                          const range<Dims> &NewRange)
+      : Id(Id), OldRange(OldRange), NewRange(NewRange) {
+    for (int i = 0; i < Dims; ++i)
+      if (Id[i] >= OldRange[i])
+        Done = true;
+  }
+
+  explicit operator bool() { return !Done; }
+
+  void updateId() {
+    for (int i = 0; i < Dims; ++i) {
+      Id[i] += NewRange[i];
+      if (Id[i] >= OldRange[i])
+        Id[i] %= OldRange[i];
+      else
+        return;
+    }
+    Done = true;
```

Comment:
This is clever, but I wonder about the performance impact of these branches and modular arithmetic.

Would it make sense to add a TODO saying we should revisit and optimize this?  I would have expected that the fastest thing to do would be to have a separate loop for each dimension, and specialize the wrapper based on the dimensionality of the kernel.  But that's more work, and I don't want to block this initial PR.

Reply:
The modular arithmetic could be removed by saving the initial ID and then doing `Id[i] = Init[i]`.

For the multidimensional case, do you mean that the fasted method would like something like this?

```c++
for (auto z = it[2]; z < OldRange[2]; z += it.get_range(2))
  for (auto y = it[1]; y < OldRange[1]; y += it.get_range(1))
    for (auto x = it[0]; x < OldRange[0]; x += it.get_range(0))
      KernelFunc({x,y,z});
```

I think it does make sense to change it like that, so I can add a todo.

Reply:
Yes, exactly.